streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -215,91 +215,54 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+        final LinkedList<Task> tasksToClose = new LinkedList<>();
+        final Set<Task> tasksToRecycle = new HashSet<>();
         final Set<Task> dirtyTasks = new HashSet<>();
 
+        // first rectify all existing tasks
         for (final Task task : tasks.values()) {
             if (activeTasks.containsKey(task.id()) && task.isActive()) {
                 updateInputPartitionsAndResume(task, activeTasks.get(task.id()));
-                if (task.commitNeeded()) {
-                    additionalTasksForCommitting.add(task);
-                }
                 activeTasksToCreate.remove(task.id());
             } else if (standbyTasks.containsKey(task.id()) && !task.isActive()) {
                 updateInputPartitionsAndResume(task, standbyTasks.get(task.id()));
                 standbyTasksToCreate.remove(task.id());
-                // check for tasks that were owned previously but have changed active/standby status
             } else if (activeTasks.containsKey(task.id()) || standbyTasks.containsKey(task.id())) {
+                // check for tasks that were owned previously but have changed active/standby status
                 tasksToRecycle.add(task);
             } else {
-                try {
-                    task.suspend();
-                    final Map<TopicPartition, OffsetAndMetadata> committableOffsets = task.prepareCommit();
-
-                    tasksToClose.add(task);
-                    if (!committableOffsets.isEmpty()) {
-                        consumedOffsetsAndMetadataPerTask.put(task.id(), committableOffsets);
-                    }
-                } catch (final RuntimeException e) {
-                    final String uncleanMessage = String.format(
-                        "Failed to close task %s cleanly. Attempting to close remaining tasks before re-throwing:",
-                        task.id());
-                    log.error(uncleanMessage, e);
-                    taskCloseExceptions.put(task.id(), e);
-                    // We've already recorded the exception (which is the point of clean).
-                    // Now, we should go ahead and complete the close because a half-closed task is no good to anyone.
-                    dirtyTasks.add(task);
-                }
+                tasksToClose.add(task);
             }
         }
 
-        if (!consumedOffsetsAndMetadataPerTask.isEmpty()) {
+        for (final Task task : tasksToClose) {
             try {
-                for (final Task task : additionalTasksForCommitting) {
-                    final Map<TopicPartition, OffsetAndMetadata> committableOffsets = task.prepareCommit();
-                    if (!committableOffsets.isEmpty()) {
-                        consumedOffsetsAndMetadataPerTask.put(task.id(), committableOffsets);
+                task.suspend(); // Should be a no-op for active tasks, unless we hit an exception during handleRevocation
```

Comment:
If we hit an exception in `handleRevocation` why would we continue here? Are we still in a "clean enough" state to actually continue?

Below we call `completeTaskCloseClean(task)` what seem incorrect for this case as it might close clean task even if we did not successfully commit before.

Reply:
If we hit an exception in `handleRevocation` on some task then we would skip out on suspending the rest of the tasks, ie the set of not-suspended tasks does not contain the task that threw (of course if one task threw an exception then its likely others will too, but not guaranteed).

But maybe it's cleaner to catch exceptions during `handleRevocation` and at least make sure every task gets suspended? I'll try that

On a related note, if we _always_  have to commit before closing (or at least attempt to), should we just remove the `writeCheckpointIfNeeded` call from `closeClean`? Seems like the `pre/postCommit` should be responsible for whether to checkpoint, not `close`. In this case, it's completely fine to _attempt_ a clean close of a dirty task, as the `closeClean` method will just maybe throw in which case we can close dirty. WDYT?

Reply:
Instead of checkpointing, we can check `if clean && commitNeeded & checkpoint != null` and then throw an exception on closeClean (which would result in calling `closeDirty`)

Reply:
I re-read the current code structure and got some questions:

1) we collect checkpoint from `prepareCommit` and check if it is not null in `postCommit`, but the actual checkpoint value itself is always collectable post the commit, and hence what's only required to that we need to know if we need to write a checkpoint file or not. Previously this needs to be decided since we may transit the state in between but now from the source code it seems to me that we would only call `prepare/post` before suspend / close ever, so this is no longer required actually, i.e. we can decide whether we need to checkpoint and then collect the checkpoint map and write the file if needed in a single call. Is that right?

2. I think I agree with you that it is cleaner to make sure in `handleRevocation`, we still transit those revoked partition's corresponding tasks to suspended even if some of their commit call failed.

Reply:
1. I think you're right, we don't need to keep track of the current `checkpoint` offsets at all and can just write the current `checkpointableOffsets` in `postCommit`
2. done

Reply:
`postCommit` only write a checkpoint for non-eos. Thus, we still need to write a checkpoint for eos case in `close()`.

Reply:
`postCommit` will always write the checkpoint if the task is in SUSPENDED, which it should always be before being closed